src/action/list_directory.rs

```diff
@@ -0,0 +1,727 @@
+    Response {
+        mode: Some(metadata.mode().into()),
+        ino: Some(metadata.ino() as u32),
+        dev: Some(metadata.dev() as u32),
+        nlink: Some(metadata.nlink() as u32),
+        uid: Some(metadata.uid() as u32),
+        gid: Some(metadata.gid() as u32),
+        size: Some(metadata.size()),
+        atime: get_accesses_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        ctime: get_status_change_time(&metadata),
+        blocks: Some(metadata.blocks() as u32),
+        blksize: Some(metadata.blksize() as u32),
+        rdev: Some(metadata.rdev() as u32),
+        flags_linux:
+        Some(get_linux_flags(file_path).unwrap_or_default() as u32),
+        symlink: if metadata.file_type().is_symlink() {
+            match fs::read_link(file_path) {
+                Ok(file) => Some(file.to_string_lossy().to_string()),
+                Err(error) => {
+                    warn!("unable to read symlink: {}", error);
+                    None
+                }
+            }
+        } else {
+            None
+        },
+        pathspec: PathSpec {
+            path_options: Some(PathOption::CaseLiteral),
+            path: file_path.clone(),
+        },
+        crtime: get_creation_time(&metadata),
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+fn fill_response(metadata: &Metadata, file_path: &PathBuf) -> Response {
+    Response {
+        size: Some(metadata.len()),
+        atime: get_accesses_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        pathspec: PathSpec {
+            path_options: Some(PathOption::CaseLiteral),
+            path: file_path.clone(),
+        },
+        crtime: get_creation_time(&metadata),
+        ..Default::default()
+    }
+}
+
+pub fn handle<S: Session>(session: &mut S, request: Request)
+                          -> session::Result<()> {
+    let dir_path = &request.pathspec.path;
+    let mut paths: Vec<PathBuf> = dir_path.read_dir()
+        .map_err(Error::ReadPath)?.filter_map(|entry| entry.ok())
+        .map(|entry| entry.path()).collect();
+    paths.sort();
+
+    for file_path in &paths {
+        let metadata = fs::symlink_metadata(file_path)
+            .map_err(Error::ReadPath)?;
+        session.reply(fill_response(&metadata, file_path))?;
+    }
+
+    Ok(())
+}
+
+/// Converts integer from proto to human-readable enum type
+fn get_enum_path_options(option: &Option<i32>) -> Option<PathOption> {
+    match option {
+        Some(poption) => match poption {
+            0 => Some(PathOption::CaseInsensitive),
+            1 => Some(PathOption::CaseLiteral),
+            2 => Some(PathOption::Recursive),
+            3 => Some(PathOption::Regex),
+            _ => None
+        },
+        _ => None,
+    }
+}
+
+fn get_path(path: &Option<String>) -> PathBuf {
+    match path {
+        Some(string_path) if !string_path.is_empty() => {
+            PathBuf::from(string_path)
+        }
```

Comment:
I think it fits in one line, no need for braces.

Reply:
It doesn't fit in 80 characters per line.

Reply:
It will if you rename `string_path` to `path` :)